### DISCUSSION TITLE
Why are my BPF's ignored?

### DISCUSSION THREAD

**Discussion by w4rc0n** [Security-Onion-Solutions/securityonion#5082] [General] [answered]
For some reason, there are multiple IP's I have explicitly set to ignore via BPF in the configs that are still alerting on nids, and being pcap'd, and imported into zeek. the `.71` address below for example:

![bpf](https://user-images.githubusercontent.com/49037260/128427174-1d8187d7-365a-469e-af05-4bf098890509.png)

![alerts](https://user-images.githubusercontent.com/49037260/128427187-027d087b-eb74-43e0-a0f8-d31c67eea104.png)

edit: I am using the global.sls option for all nodes.

**Comment by weslambert** [comment, marked as answer]
Have you followed all of the instructions here?
https://docs.securityonion.net/en/latest/bpf.html?highlight=bpf#global-bpf

**Reply by w4rc0n** [reply]
Yes.

**Reply by petiepooo** [reply]
Is it because '&' is a reserved YAML character and the line is not quoted?  They are not quoted in the example in the page linked above, but it also says above the example that '&' is reserved and should thus be quoted...

**Reply by w4rc0n** [reply]
Worth a shot I suppose

**Reply by w4rc0n** [reply]
Well now it's just filtering everything it seems.

**Reply by w4rc0n** [reply]
All traffic to suricata is being filtered out, and chunks of zeek and pcap data just being filtered out

**Reply by weslambert** [reply]
You might need some `()`s so that the `and dst`s aren't considered as inclusive filters.

**Reply by w4rc0n** [reply]
That was it, thank you @weslambert

**Reply by thegreekman** [reply]
@w4rc0n If you do not mind, what is the syntax of your last bpf filter with the src and dst? I can't seem to get mine to work properly.

**Reply by w4rc0n** [reply]
they look like

`  - (not host 192.168.34.2 and port 333)`

**Reply by thegreekman** [reply]
Thanks, and what about the one with src and dst in it? 

Im trying to do something like this:

- not (src host 192.xx.xx.xx and not dst host 192.xx.xx.xx)

**Comment by w4rc0n** [comment]
You could try something like:


   -  "!(src host 192.xx.xx.xx and not dst host 192.xx.xx.xx)"

**Reply by thegreekman** [reply]
When I input that, traffic is still coming from the source and destination IP  I entered. Do you know why this happening?

**Reply by petiepooo** [reply]
Let's break that filter string down.  Inside the parentheses, you're looking for traffic with a specific src IP but not to the destination IP.  Then you're negating everything outside the parentheses.  So then you should see everything except that specific scenario.  Note that that includes traffic between the two IPs, which is what you appear not to want.

If you want to exclude traffic going from A to B, you'll want to first positively identify the traffic to exclude by endpoint IPs, then exclude it by wrapping in parens and negating it with 'not' (or '!').  If you want to also exclude the return traffic (from destination to source), drop the src and dst qualifiers and just use host.

Also, if you're never expecting VLAN tagging, I suggest specifying 'ip' on all your filters; without that, it will check for a vlan tag, then branch to two different checks based on whether it sees the tag or not.  Hint: use the -d option for tcpdump to test your expressions and see the difference.

Assuming your true intent is to drop ALL traffic between those two hosts, I think the filter you're looking for is
> ! ( ip host 1.1.1.24 && host 1.1.1.208 )

And don't forget to add the ending 'and' (or '&&') to all except the final entry.

**Reply by thegreekman** [reply]
I changed the condition to ! (ip host 192.168.1.1 && host 192.168.1.40), but unfortunately the traffic is still coming through. My true intent is to drop all traffic between the two hosts like your final statement said.

**Comment by w4rc0n** [comment]
Is that your only BPF rule in place? If not, can we see what you have?

**Reply by thegreekman** [reply]
This is my current setup.  The not port rules work fine. If I insert a not rule on a single host it works as well, but when I try to combine src and dst it doesn't seem to do anything.  

![image](https://user-images.githubusercontent.com/27656903/131540458-74f37264-22c7-4694-a2c0-43cddf692334.png)

**Reply by w4rc0n** [reply]
So, let's say your filter is this: 

`src host 192.168.1.1 and not dst host 192.168.1.22`

The filter says: I want to see everything where 192.168.1.1 is the source, except not when the destination is 192.168.1.22.

But then you're inverting that with `!`. So you're saying, I want to exclude everything where the source host is 192.168.1.1 where the destination isn't 192.168.1.22. I don't think that's really what you want to do right?

**Reply by thegreekman** [reply]
Yeah you are right, i had the wrong condition set. I changed it to ! (ip host 192.168.1 && host 192.168.1.45). I want to block all traffic that originates from .1 and ends at .45. Unfortunately, even with this, it still shows traffic between these 2 addresses

**Reply by w4rc0n** [reply]
Do you know if the traffic you're analyzing is VLAN tagged?

**Reply by thegreekman** [reply]
It is not vlan tagged

**Reply by petiepooo** [reply]
Some suggestions:

- Standardize on 'not' instead of '!' and 'and' instead of '&&' just to help eliminate parsing issues with special characters (other than parentheses).
- build a full bpf with all line contents concatenated, eg: 'not port 53 and not port 80 and not port 443 and not ( ip host 192.168.1.1 and host 192.168.1.45 )' and feed that to tcpdump -d to see if it parses and compiles like you expected.
- Make a trivial change (like reordering two of the port expressions), then run 'sudo salt-call state.highstate' and review the output for errors (you should see your change being picked up).

**Reply by petiepooo** [reply]
another small optimization: specify 'not tcp port 80' (also for 443) otherwise it checks both tcp and udp.  DNS can fallback to tcp, though, so you may want to leave port 53 as is.
